Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs

```diff
@@ -0,0 +1,150 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer
+{
+    using System;
+    using System.Collections.Generic;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+    using Microsoft.ApplicationInsights.WindowsServer.Implementation;
+
+    /// <summary>
+    /// Provides default values for the heartbeat feature of Application Insights that
+    /// are specific to Azure App Services (Web Apps, Functions, etc...).
+    /// </summary>
+    internal class AppServicesHeartbeatTelemetryModule : ITelemetryModule, IDisposable
+    {
+        internal const int SiteNameMaxLength = 256;
+        internal const int StampNameMaxLength = 256;
+        internal const int HostNameMaxLength = 256;
+
+        /// <summary>
+        /// Environment variables and the Application Insights heartbeat field names that accompany them.
+        /// </summary>
+        internal static readonly KeyValuePair<string, string>[] WebHeartbeatPropertyNameEnvVarMap = new KeyValuePair<string, string>[]
+        {
+            new KeyValuePair<string, string>("appSrv_SiteName", "WEBSITE_SITE_NAME"),
+            new KeyValuePair<string, string>("appSrv_wsStamp", "WEBSITE_HOME_STAMPNAME"),
+            new KeyValuePair<string, string>("appSrv_wsHost", "WEBSITE_HOSTNAME")
+        };
+
+        private static AppServicesHeartbeatTelemetryModule instance;
+        private object lockObject = new object();
+        private bool isInitialized = false;
+
+        /// <summary>
+        /// Initializes a new instance of the<see cref="AppServicesHeartbeatTelemetryModule" /> class.
+        /// </summary>
+        internal AppServicesHeartbeatTelemetryModule()
+        {
+            AppServicesHeartbeatTelemetryModule.Instance = this;
+        }
+
+        /// <summary>
+        /// Gets a value to provide internal access to the only instance of this class that *should* be available.
```

Comment:
I don't understand why it is a singleton. Can you provide more details

Reply:
We need a way to allow for the update of the heartbeat values associated to AppServices. I made this a singleton to allow other modules to signal this module to update the values. Perhaps there is a better mechanism for this, but I am not aware of one at this time. (I have thoughts on how to improve it in the future though, but it might take some time).